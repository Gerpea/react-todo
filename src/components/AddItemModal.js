import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { Col, Form, Modal, Button } from 'react-bootstrap'

const AddItemModal = ({ todos, onCreate, ...rest }) => {
  const [parents, setParents] = useState([])
  const [parent, setParent] = useState()
  const [title, setTitle] = useState('')

  useEffect(() => {
    setParents([
      { title: 'Не выбрано' },
      ...todos.filter((todo) => getParentCount(todos, todo.uid) < 2),
    ])
  }, [todos])

  return (
    <Modal {...rest} animation={false} centered>
      <Modal.Header>
        <Modal.Title>Добавить элемент</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form
          onSubmit={(e) => {
            e.preventDefault()
          }}>
          <Form.Group controlId='parent-todo' data-testid='select-parent'>
            <Form.Row>
              <Col>
                <Form.Label>Родительский элемент</Form.Label>
              </Col>
              <Col>
                <Form.Control
                  as='select'
                  value={parent}
                  onChange={(e) => {
                    setParent(e.target.value)
                  }}>
                  {parents.map((todo, i) => {
                    return (
                      <option value={todo.uid} key={todo.uid ?? i}>
                        {todo.title}
                      </option>
                    )
                  })}
                </Form.Control>
              </Col>
            </Form.Row>
          </Form.Group>
          <Form.Group controlId='title'>
            <Form.Row>
              <Col>
                <Form.Label>Заголовок</Form.Label>
              </Col>
              <Col>
                <Form.Control
                  type="input'"
                  defaultValue=''
                  onChange={(e) => setTitle(e.target.value)}
                />
              </Col>
            </Form.Row>
          </Form.Group>
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <Button
          variant='primary'
          className='w-100'
          onClick={() => {
            rest.onHide?.call()
            onCreate?.call(undefined, title, parent)
            setParent(undefined)
          }}
          data-testid='add-todo'>
          Добавить
        </Button>
      </Modal.Footer>
    </Modal>
  )
}

function getParentCount(todos, uid) {
  const parentUid = todos.find((todo) => todo.uid === uid)?.parentUid
  if (parentUid) {
    return 1 + getParentCount(todos, parentUid)
  } else {
    return 0
  }
}

AddItemModal.propTypes = {
  todos: PropTypes.array,
  onCreate: PropTypes.func,
}

AddItemModal.defaultProps = {
  todos: [],
  onCreate: () => {},
}

export default AddItemModal
