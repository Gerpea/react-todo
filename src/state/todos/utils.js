function removeTodo(state, uid) {
  const children = getChildren(state, uid)

  let withoutChild = state
  children.forEach((child) => {
    withoutChild = removeTodo(state, child.uid)
  })

  return withoutChild.reduce((acc, cur) => {
    if (cur.uid !== uid) {
      acc.push(cur)
    }
    return acc
  }, [])
}

function getChildren(state, uid) {
  return state.filter((item) => item.parentUid === uid)
}

function getParentCount(state, uid) {
  const parentUid = state.find((item) => item.uid === uid)?.parentUid
  if (parentUid) {
    return 1 + getParentCount(state, parentUid)
  } else {
    return 0
  }
}

export { removeTodo, getChildren, getParentCount }
