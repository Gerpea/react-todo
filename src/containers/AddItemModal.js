import { connect } from 'react-redux'

import { addTodo } from '../state/todos/actions'

import AddItemModal from '../components/AddItemModal'

const mapStateToProps = (state) => ({
  todos: state.todos,
})

const mapDispatchToProps = {
  onCreate: (title, parentUid) => addTodo(title, parentUid),
}

export default connect(mapStateToProps, mapDispatchToProps)(AddItemModal)
