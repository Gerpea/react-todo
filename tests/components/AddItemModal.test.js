import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { emptyWrapper, render, cleanup, fireEvent } from '../utils'
import AddItemModal from '../../src/components/AddItemModal'

afterEach(cleanup)

it('renders without crashing', () => {
  render(<AddItemModal />)
})

describe('todos', () => {
  it('should render default when todos is empty', () => {
    const { getByTestId } = render(<AddItemModal show={true} />)

    expect(getByTestId('select-parent').getElementsByTagName('select')[0].childElementCount).toBe(1)
  })

  it('should render correct number of parents', () => {
    const todos = [{}, {}, {}]

    const { getByTestId } = render(<AddItemModal show={true} todos={todos} />)

    expect(getByTestId('select-parent').getElementsByTagName('select')[0].childElementCount).toBe(
      todos.length + 1
    )
  })

  it('should not render parents which is more than 2th generation', () => {
    const todos = [
      { title: 'Item1', uid: '1' },
      { title: 'Item2', uid: '2', parentUid: '1' },
      { title: 'Item3', uid: '3', parentUid: '2' },
    ]

    const { getByTestId } = render(<AddItemModal show={true} todos={todos} />)

    expect(getByTestId('select-parent').getElementsByTagName('select')[0].childElementCount).toBe(
      todos.length
    )
  })
})

describe('onCreate', () => {
  it('should be called when add button clicked', () => {
    const onCreate = jest.fn()
    const leftButton = { button: 0 }
    const { getByTestId } = render(<AddItemModal show={true} onCreate={onCreate} />)

    fireEvent.click(getByTestId('add-todo'), leftButton)

    expect(onCreate.mock.calls.length).toBe(1)
  })
})

ReactDOM.createPortal = (node) => node
it('matches snapshot', () => {
  const todos = [
    { title: 'Item 1', uid: '1' },
    { title: 'Item 2', uid: '2' },
  ]
  const container = document.createElement('div')

  const tree = renderer
    .create(
      <div>
        {ReactDOM.createPortal(emptyWrapper(<AddItemModal show={true} todos={todos} />, container))}
      </div>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
