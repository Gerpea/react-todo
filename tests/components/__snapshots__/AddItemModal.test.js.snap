// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    className="modal-backdrop show"
    onClick={[Function]}
  />
  <div
    aria-modal={true}
    className="modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex={-1}
  >
    <div
      className="modal-dialog modal-dialog-centered"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <div
            className="modal-title h4"
          >
            Добавить элемент
          </div>
        </div>
        <div
          className="modal-body"
        >
          <form
            className=""
          >
            <div
              className="form-group"
              data-testid="select-parent"
            >
              <div
                className="form-row"
              >
                <div
                  className="col"
                >
                  <label
                    className="form-label"
                    htmlFor="parent-todo"
                  >
                    Родительский элемент
                  </label>
                </div>
                <div
                  className="col"
                >
                  <select
                    className="form-control"
                    id="parent-todo"
                    onChange={[Function]}
                  >
                    <option>
                      Не выбрано
                    </option>
                    <option
                      value="1"
                    >
                      Item 1
                    </option>
                    <option
                      value="2"
                    >
                      Item 2
                    </option>
                  </select>
                </div>
              </div>
            </div>
            <div
              className="form-group"
            >
              <div
                className="form-row"
              >
                <div
                  className="col"
                >
                  <label
                    className="form-label"
                    htmlFor="title"
                  >
                    Заголовок
                  </label>
                </div>
                <div
                  className="col"
                >
                  <input
                    className="form-control"
                    defaultValue=""
                    id="title"
                    onChange={[Function]}
                    type="input'"
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="w-100 btn btn-primary"
            data-testid="add-todo"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Добавить
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
